src/sender.js

```diff
@@ -0,0 +1,114 @@
+require('dotenv').config()
+const path = require('path')
+const Web3 = require('web3')
+const { connectSenderToQueue } = require('./services/amqpClient')
+const { redis, redlock } = require('./services/redisClient')
+const { getGasPrices } = require('./services/gasPrice')
+const { sendTx, sendRawTx } = require('./tx/sendTx')
+const { getNonce } = require('./tx/web3')
+const { syncForEach } = require('./utils/utils')
+
+const { VALIDATOR_ADDRESS, VALIDATOR_ADDRESS_PRIVATE_KEY } = process.env
+
+const config = require(path.join('../config/', process.argv[2]))
+
+const provider = new Web3.providers.HttpProvider(config.url)
+const web3Instance = new Web3(provider)
+const nonceLock = `lock:${config.id}:nonce`
+const nonceKey = `${config.id}:nonce`
+let chainId = 0
+
+async function initialize() {
+  chainId = await sendRawTx({
+    url: config.url,
+    params: [],
+    method: 'net_version'
+  })
+  connectSenderToQueue({
+    queueName: config.queue,
+    cb: main
+  })
+}
+
+async function readNonce(forceUpdate) {
+  if (forceUpdate) {
+    return getNonce(web3Instance, VALIDATOR_ADDRESS)
+  }
+
+  const result = await redis.get(nonceKey)
+  return result ? Number(result) : getNonce(web3Instance, VALIDATOR_ADDRESS)
+}
+
+function updateNonce(nonce) {
+  redis.set(nonceKey, nonce)
+}
+
+async function main({ msg, ackMsg, nackMsg, sendToQueue }) {
+  if (redis.status !== 'ready') {
+    console.log('Redis not connected.')
+    await setTimeout(() => nackMsg(msg), 5000)
```

Comment:
Two questions here:

1. Why the timeout instead of just immediately nacking the message?
2. Why the await? `setTimeout` doesn't return a promise.

Reply:
You are right, I'll fix that line into `nackMsg(msg)`